internal/contour/resources.go

```diff
@@ -17,29 +17,30 @@
 package contour
 
 import (
+	xds "github.com/envoyproxy/go-control-plane/pkg/cache/types"
 	"github.com/projectcontour/contour/internal/dag"
-	"github.com/projectcontour/contour/internal/xds"
+	contourxds "github.com/projectcontour/contour/internal/xds"
 )
 
 // ResourceCache is a store of an xDS resource type. It is able to
 // visit the dag.DAG to update the its resource collection, then
 // serve those resources over xDS.
 type ResourceCache interface {
 	dag.Observer
-	xds.Resource
+	contourxds.Resource
 }
 
 // ResourcesOf transliterates a slice of ResourceCache into a slice of xds.Resource.
-func ResourcesOf(in []ResourceCache) []xds.Resource {
-	out := make([]xds.Resource, len(in))
+func ResourcesOf(in map[xds.ResponseType]ResourceCache) []contourxds.Resource {
```

Comment:
I don't think you'll need to do this if you can make the mapping internal to the snapshot observer.

Reply:
I was wanting to re-use the ResourceCache, but need to know the types later on of which is the listener, which is endpoints, etc. 

Can you explain a bit more on your thought?

Reply:
What I was thinking was using `ResourceCache.TypeURL()` to internally map to the control-plane `ResponseType`. The type URLs are global unique, so you can just map internally and keep the implementation details that need `ResponseType` private.

Reply:
I can do that and remove the map, but I'd rather have a quick lookup to the resources than have to loop through them all and figure out what type is which. 

What is the downside of switching the type to a map?

If we ditch the map I'd end up with something like this in the snapshot constructor which to me feels bad, I'd rather just lookup the index of the map:

```
	for _, r := range resources {
		switch r.TypeURL() {
		case resource.ClusterType:
			clusterCache = r
		case resource.RouteType:
			routeCache = r
		case resource.ListenerType:
			listenerCache = r
		case resource.SecretType:
			secretCache = r
		case resource.EndpointType:
			endpointsCache = r
		}
	}
```

Reply:
Both the integer types and the typeURLs are well-known constants with a permanently fixed relationship, so you can preinitialize a constant map:
```Go
m := map[string]xds.ResponseType {
resource.EndpointType: types.Endpoint,
...
}

snapshot[m[r.TypeURL()]] = resourcesOf(r.Contents())
```

The benefit of this approach is that knowledge of the go-control-plane snapshot types can be encapsulated by the one component that really needs to know. The rest of Contour can be oblivious.

Reply:
It's still a map though, so why are strings better than types? I'd prefer the types since they are easier to work with unless I'm not following.

Reply:
What I'm trying to avoid is the need for the go-control-plane types to propagate over the whole codebase. Only the snapshot handler needs them and I'm suggesting that we can localize their use internally to that.

Reply:
The only way to hide from the other bits is to remove the map as described in https://github.com/projectcontour/contour/pull/2850#discussion_r480177395. 

I still prefer the type as implemented keeping the map.

Reply:
To keep this moving I did what you said @jpeach, I don't want this to block hold up the implementation.